feat(Geometry/Manifold): orientable manifolds

Rida-Hamadani opened the pull request:
This module defines orientable manifolds. [This](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Orientable.20Surface) Zulip topic is related.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Rida-Hamadani committed:
feat: orientable manifolds

Rida-Hamadani committed:
docs

Rida-Hamadani committed:
useless indentation

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,80 @@
+Jacobian is strictly positive on that set.
+-/
+def OrientationPreserving (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s,
```
Also: I think you can remove the differentiability requirement, as `fderiv` has junk value `0` whenever the map is not differentiable. (Perhaps add a comment to that effect, if you find one helpful.)
That ought to simplify later proof, right?

Rida-Hamadani replied:
I've considered doing that, and it turns out that I will need to add another condition to get differentiability from the positivity of `fderiv`, one possible condition to obtain that is `FiniteDimensional.finrank ℝ H ≠ 0` (cf. `LinearMap.det_zero`). Since I need to add a condition anyways, I decided to go with the one that sounded to me closer to spoken mathematics. Do you think that the `FiniteDimensional.finrank ℝ H ≠ 0` condition is better in our case?

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,80 @@
+    h x <| Set.mem_inter hxu hxv
+  congr {f g u} hu fg hf := by
+    intro x hx
+    have := Filter.eventuallyEq_of_mem (IsOpen.mem_nhds hu hx) fg
```
This line and the next ones look like they could use dot notation, right?

Rida-Hamadani replied:
Dot notation works for the next lines, but not for this one (the `have` proof terms).

grunweg reviewed (commented):
Thanks for this PR, I am excited to see somebody work on it! I have one comment, which could simplify your definition and proofs, and one follow-up request.

Right now, this is just a definition, without any API. Do you mind adding some lemmas about this definition, such as
- the identity is orientation-preserving
- the composition of two orientation-preserving maps is orientation-preserving
- (optional) define orientation-reversing maps; show that (over a connected set), each diffeomorphism is either orientation-preserving or reversing
- (optional) the composition of or.-preserving and -reversing is orientation-reversing; composing two orientation-reversing maps yields something orientation-preserving

Another future possibility (perhaps not for this PR) could be to define the groupoid of smooth orientation-preserving manifolds. You need not do all of these! But the first two items above would be nice, to confirm this definition is usable.

Rida-Hamadani committed:
apply style suggestions from grunweg

grunweg commented:
Another thought just came to my mind: there is a notion of orientability of *topological* manifolds, using local orientations. (This is not an issue for surfaces, as topological and smooth manifolds are the same in dimension less than four.)
Having this definition is useful now, but it's a definition only for differentiable manifolds. Can you add a comment clarifying this assumption?

Rida-Hamadani committed:
add note about differentiable vs topological manifolds orientability

grunweg commented:
Another interesting follow-up lemma could be to define orientation-preserving (or reversing) map between smooth manifolds. Up to you if you'd like to do this!

Rida-Hamadani committed:
OrientationReversing

Rida-Hamadani committed:
switch names

Rida-Hamadani committed:
add `OrientableSmoothManifold`

Rida-Hamadani committed:
shake

Rida-Hamadani committed:
doc-string

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+
+## Main Definitions
+
+- `OrientationPreserving` : a map is orientation-preserving on a given set if the determinant of
```
Let's be more precise: "a map between normed spaces ..."
(This notion also exists for maps between oriented manifolds, which this PR is a step towards.)

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+
+- `OrientationPreserving` : a map is orientation-preserving on a given set if the determinant of
+  its Jacobian is strictly positive on that set.
+- `orientationPreservingGroupoid` : the groupoid of partial homeos of `H` which are
```
Can you mention the corresponding declarations for orientation-reversing as well?

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+- `orientationPreservingGroupoid` : the groupoid of partial homeos of `H` which are
+  orientation-preserving.
+- `OrientableManifold`: a type class saying that the charted space `M`, modelled on the space `H`,
+  admis an orientation.
```
```suggestion
  admits an orientation.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+section OrientationPreserving
+
+/--
+A map is orientation-preserving on a given set if it is differentiable and the determinant of its
```
Similarly as above, can you clarify this to "A map between normed spaces"? Same comment for the other declarations below.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+-/
+def OrientationPreserving (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, 0 < (fderiv ℝ f x).det
+
```
If I understand Lean's convention (0^0 = 1) correctly, this means 0-dimensional manifolds are orientable. (This is mathematically correct: these are simply points; an orientation of a zero-manifold simply assigns each point a value $$\pm 1$$.)
Can you add a lemma stating this, that `H` being 0-dimensional implies f is orientation-preserving?

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+def OrientationReversing (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, (fderiv ℝ f x).det < 0
+
+lemma OrientationPreserving.DifferentiableAt [Nontrivial H] [FiniteDimensional ℝ H] {f : H → H}
```
Three small comments: all of them also apply to the other lemmas below.
- I think you don't need the `Nontrivial H` assumption: if `H` is zero-dimensional, `f` is always differentiable. I think mathlib has this already; ping me if you get stuck finding it!
- By default, I would put the `x` "left of the colon", i.e. write "(x : s) : DifferentiableAt \R f x" instead. Do you have a reason to do this differently?
- In terms of naming, I'd write "OrientationPreserving.differentiableAt".

Rida-Hamadani replied:
Thank you for pointing these out! Regarding point two: instead of writing `(x : s)`, I decided to look at similar lemmas in mathlib and imitate how it is done there (e.g. `DifferentiableWithinAt.differentiableAt`).

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+  rw [(fderiv ℝ g (f x)).toLinearMap.det_comp (fderiv ℝ f x).toLinearMap]
+  exact mul_pos (hg (f x) hxv) (hf x hxu)
+
+lemma orientationReversing_comp_OrientationPreserving [Nontrivial H] [FiniteDimensional ℝ H]
```
`orientationReversing_comp_orientationPreserving`, please - thanks!

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+    [FiniteDimensional ℝ H] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    [OrientableManifold H M] (I : ModelWithCorners ℝ H M) [SmoothManifoldWithCorners I M] extends
+  SmoothManifoldWithCorners I M ∧ OrientableManifold H M : Prop
+
```
Can you add the corollary that 0-dimensional manifolds are always orientable?

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+
+## TODO
+
+- Define orientable `0`-manifolds.
```
IIUC, this PR already handles the 0-dimensional case (and correctly). Instead, I would extend this TODO slightly differently: I'll leave this blank for now and try to return here after dinner.

grunweg replied:
So, here's an initial list of what I'd write.
- try to generalise this discussion to any non-trivially normed field
- on a given connected set, a diffeomorphism is either orientation-preserving or orientation-reversing
- a normed space (with the trivial model) is orientable
- the n-sphere is orientable
- products of orientable manifolds are orientable

- define orientations of a smooth manifold, and show that a manifold is orientable if and only if it admits an orientation
- define orientation-preserving and reserving maps between manifolds

grunweg reviewed (commented):
Thanks for addressing these comments so quickly! I think this PR is converging quickly. I took a quick look; all comments I have are pretty minor. (I didn't look at the proofs in detail yet.) It's time for dinner now; I'll try my best to return to this afterwards.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+/-!
+# Orientable Manifolds
+
+This module defines orientable manifolds.
```
Here's a slightly longer version, if you prefer:
```This file defines orientable manifolds: a differentiable manifold is orientable if and only if it admits an orientable atlas, i.e. an atlas whose transition functions have positive Jacobian.```

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+## TODO
+
+- Define orientable `0`-manifolds.
+
```
Thinking about it more, it would be nice to mention the choice of definition also. What do you think about this (rewrapping lines as needed)?
```suggestion
## Implementation notes

The current definitions work for differentiable manifolds (and is simpler to define this way).
For topological manifolds, orientability can be defined using *local* orientations
(which mathlib cannot even state yet, as there is no e.g. singular homology).
In the future, it would be nice to generalise these definitions to allow for topological manifolds also, and relate them to the current definition.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,135 @@
+class OrientableSmoothManifold (H : Type*) [Nontrivial H] [NormedAddCommGroup H] [NormedSpace ℝ H]
+    [FiniteDimensional ℝ H] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    [OrientableManifold H M] (I : ModelWithCorners ℝ H M) [SmoothManifoldWithCorners I M] extends
+  SmoothManifoldWithCorners I M ∧ OrientableManifold H M : Prop
```
Hm, I'd need to think if this does what we'd want: this definition asks for "M has a smooth atlas, and M has an orientable atlas" - whereas the obviously correct definition would be "M has a smooth orientable atlas" (i.e., one atlas which is both smooth and orientable).

grunweg reviewed (commented):
Here are my remaining comments for today.

Rida-Hamadani committed:
grunweg's suggestions -- part 1

Rida-Hamadani committed:
grunweg's suggestions -- part 2 -- remove `Nontrivial H` hypotheses

Rida-Hamadani committed:
grunweg's suggestions -- part 3 -- zero dimensions

Rida-Hamadani committed:
turn instance to lemma

Rida-Hamadani committed:
remove forall x

Rida-Hamadani committed:
`DifferentiableAt` to `differentiableAt`

Rida-Hamadani committed:
OrientableSmoothManifold

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+  and smooth.
+
+## Main Results
+
```
What do you think about mentioning that 0-dimensional manifolds are always orientable?

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+
+## TODO
+
+- Generalize this discussion to any non-trivially normed field.
```
I tried this and got stuck somewhere; let's tone down the wording.
```suggestion
- Generalize this discussion to other fields; it certainly makes sense over ℚ

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+- Products of orientable manifolds are orientable.
+- Define orientations of a smooth manifold, and show that a manifold is orientable if and only if it
+  admits an orientation.
+- Define orientation preserving and reserving maps between manifolds.
```
```suggestion
- Define orientation preserving and reversing maps between manifolds.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+
+lemma orientationPreserving_of_zero_dim (f : H → H) (s : Set H) (h : Module.rank ℝ H = 0) :
+    OrientationPreserving f s :=
+  fun _ _ ↦
```
You can golf this proof, using `LinearMap.det_eq_one_of_finrank_eq_zero.`
(I am not sure which way of saying "is zero-dimension" is the best one; a maintainer can certainly weigh in on this.)
```
lemma orientationPreserving_of_zero_dim (f : H → H) (s : Set H)
    (h : FiniteDimensional.finrank ℝ H = 0) : OrientationPreserving f s := by
  intro _ _
  simp [LinearMap.det_eq_one_of_finrank_eq_zero h]
```

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+
+end OrientationPreserving
+
+section OrientableManifold
```
I personally prefer the comment on top of the section name. I wonder if the current behaviour is also incidental due to the port. Not a strong opinion, in any way.
```
 /-! ### Orientable manifolds -/

section OrientableManifold
```

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+lemma orientableManifold_of_zero_dim (H : Type*) [NormedAddCommGroup H] [NormedSpace ℝ H]
+    [FiniteDimensional ℝ H] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    (h : Module.rank ℝ H = 0) : OrientableManifold H M where
+  compatible := fun {_ _} _ _ ↦
```
```suggestion
  compatible := fun _ _ ↦

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+lemma OrientationPreserving.differentiableAt [FiniteDimensional ℝ H] {f : H → H} {s : Set H}
+    (h : OrientationPreserving f s) {x : H} (hs : x ∈ s) : DifferentiableAt ℝ f x := by
+  cases subsingleton_or_nontrivial H
+  · apply DifferentiableWithinAt.differentiableAt
```
The first case can be golfed a little, to
```
apply ((s.subsingleton_of_subsingleton).differentiableOn _ hs).differentiableAt
exact mem_nhds_discrete.mpr hs
```
I would argue this is mathematically "trivial", hence fine.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+    exact s.subsingleton_of_subsingleton
+    exact hs
+    exact mem_nhds_discrete.mpr hs
+  · unfold OrientationPreserving at h
```
In general, `unfold` is a bit of a "code smell": a sign that something is not ideal (such as, a missing simple lemma about the definition to use). In this case, you can just `rw` by the definition instead.
```suggestion
  · rw [OrientationPreserving] at h

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+
+lemma OrientationReversing.differentiableAt {f : H → H} {s : Set H} (h : OrientationReversing f s)
+    {x : H} (hs : x ∈ s) : DifferentiableAt ℝ f x := by
+  unfold OrientationReversing at h
```
Please also `rw` here.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+    (hf : OrientationPreserving f u) (hg : OrientationPreserving g v) :
+    OrientationPreserving (g ∘ f) (u ∩ f ⁻¹' v) := by
+  intro x ⟨hxu, hxv⟩
+  rw [fderiv.comp x (hg.differentiableAt hxv) (hf.differentiableAt hxu)]
```
Here is a golf of the last three lines ("unfold" to "exact").
- avoiding `unfold` in favour of `rw`
- in fact, `simp` can also be used (which makes this even shorter)
- in the last step, `simp` followed by `exact` can often be converged to `simpa [...] using`: I think that is more readable here.
Overall, I'd suggest
```
simpa [ContinuousLinearMap.det] using mul_pos (hg (f x) hxv) (hf x hxu)
```

Please make the same change to the other three lemmas below.

grunweg commented on Mathlib/Geometry/Manifold/Orientable.lean:
```diff
@@ -0,0 +1,192 @@
+class OrientableSmoothManifold (H : Type*) [NormedAddCommGroup H] [NormedSpace ℝ H]
+    [FiniteDimensional ℝ H] {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] (M : Type*)
+    [TopologicalSpace M] [ChartedSpace H M] (I : ModelWithCorners ℝ E H) extends
+  HasGroupoid M ((contDiffGroupoid ⊤ I) ⊓ orientationPreservingGroupoid) : Prop
```
You *could* extract that groupoid and give it a name. Up to you!

grunweg replied:
I think this is fine as-is; if necessary, it can always be extracted later.

grunweg reviewed (commented):
Wow, you're quick! I took a close look at the proofs and have a few golfs to suggest. Should be quick to address. After that, it's probably ready to be maintainer merged.

Rida-Hamadani committed:
golfs by grunweg

grunweg commented:
Thanks! Let's have another pair of eyes on this.
maintainer merge

ocfnash commented:
I looks like this only deals with the case of manifolds without boundary. I think this approach to orientability can be made to work for boundary (though it's a bit fiddly).

What if I wanted to use this definition state Stokes's theorem?

For the sake of a concrete goal, I'd be much happier merging something that included an instance of orientability for `(Icc (0 : ℝ) 1)`.

grunweg committed:
Fix the definition of orientable manifolds.

grunweg committed:
chore: extract contDiffOrientationPreservingGroupoid

grunweg committed:
Show groupoid properties, part 1.
This pointed out that the second condition was in fact necessary;
adapted some of the proofs to it, but not all of them.

Rida-Hamadani committed:
remove two `sorry`s

grunweg committed:
Golf.

grunweg committed:
Speed up.

grunweg committed:
One more sorry.

grunweg committed:
A fin-dim normed space is orientable

Rida-Hamadani committed:
remove 1 `sorry`

Rida-Hamadani committed:
remove useless comment, i need to golf and tidy more later ...

Rida-Hamadani committed:
remove a `have`

Rida-Hamadani committed:
complete the proof of `orientationPreservingPregroupoid`

Rida-Hamadani committed:
last sorry!

Rida-Hamadani committed:
fix comment

Rida-Hamadani committed:
use order similar to `SmoothManifoldWithCorners`

Rida-Hamadani committed:
work on `Icc_orientable_manifold`

grunweg committed:
Easy clarifications.

grunweg committed:
Reduce to a computation in one real variable.

Rida-Hamadani committed:
Merge branch 'master' into rida/orientableManifolds

Rida-Hamadani committed:
maintain and deal with bitrot

Rida-Hamadani committed:
more maintenance

Rida-Hamadani committed:
choice of a sign to `HalfEuclideanSpace`

Rida-Hamadani committed:
ill brb

Rida-Hamadani committed:
Revert "choice of a sign to `HalfEuclideanSpace`"

This reverts commit 959e19b2b3e1f28a8f552f10d4e33421911bf164.

Rida-Hamadani committed:
init `IsOrientedManifold` and `IsOrientedSmoothManifold` classes

Rida-Hamadani committed:
yaomd (yet another orientable manifold definition) (maybe has a mistake but will look into it more)

Rida-Hamadani committed:
highlighting the new challenges
